Azure Ls v2 local disk setup

tarzanek opened the pull request:
fixes #8325

The iotune tests happened on Centos 8.2 both with stock and elrepo kernel, using Scylla 4.3 rc3

results are in https://docs.google.com/spreadsheets/d/1_uYq8UxY47XF5jreetrpleykLPqNGjfPXIirvTPh6rk/edit#gid=1101336711

tarzanek commented on dist/common/scripts/scylla_util.py:
```diff
@@ -343,6 +343,183 @@ def user_data(self):
+
+    def is_supported_instance(self):
+        """Returns if this instance type belongs to supported ones for nvmes"""
+        if self.instance_class() in ['L8s', 'L16s', 'L32s' ,'L48s','L64s','L80s']:
```
this could optionally use the keyset of instanceToDiskCount

tarzanek replied:
fixed

tarzanek commented:
@bentsi could you have a look please?

slivne commented:
I don't understand how the numbers align with the official specs MSFT Azure https://docs.microsoft.com/en-us/azure/virtual-machines/lsv2-series posted - but reviewers will ask the questions I am sure

tarzanek commented:
@slivne bursting applies to all other disks than nvme, nvme has a column there in above link, I don't think bursting is used for it
Azure VMs unlike other VMs always get a data folder attached with persistent storage and I'd expect bursting is used there and for OS disk.
see https://docs.microsoft.com/en-us/azure/virtual-machines/disk-bursting
and
https://azure.microsoft.com/en-us/blog/better-performance-with-bursting-enhancement-on-azure-disks/
none of them talks about nvmes

tarzanek commented:
also if you check the google doc, our tests for reads were always a bit higher, so I've set them to what is the read SLA in MS specs
for write I guess we will need to go with smallest of my few attempts
I tried diskxplorer one one instance too and it matched the expected number of queues (defaults that scylla assigns based on cpus), so I didn't run it and I rely just on 4.3 iotune(with Pavels fixes) for write iops and bandwidth

bentsi commented on dist/common/scripts/scylla_util.py:
```diff
@@ -343,6 +343,183 @@ def user_data(self):
+    @staticmethod
+    def is_azure_instance():
+        """Check if it's Azure instance via DNS lookup to metadata server."""
+        import socket
```
imports should be on top of the file

bentsi commented on dist/common/scripts/scylla_util.py:
```diff
@@ -343,6 +343,183 @@ def user_data(self):
+        self.__osDisks = None
+
+    @staticmethod
+    def is_azure_instance():
```
looks like a copy-paste from `is_gce_instance` with small changes, generalize by moving the code into separate function or create a base class that all `*_instance` classes will inherit from.

tarzanek replied:
refactoring will happen in next PR

avikivity commented:
I'm not reviewing a commit message without an explanation. Phrase the subject as an action done to the code, not a random noun.

tarzanek commented:
rebased, split commits, tried to write the message in form what was done to code
@avikivity can you review if you will have a spare minute?

@bentsi I'd like to refactor all 3 classes we have (aws, gcp, azure) in a separate commit, I don't want the refactoring to be part of this commit if possible

yaronkaikov commented on dist/common/scripts/scylla_io_setup:
```diff
@@ -277,7 +277,54 @@ if __name__ == "__main__":
+                logging.error('This is not a recommended Google Cloud instance setup for auto local disk tuning, running manual iotune.')
+                run_iotune()
+        elif azure_instance().is_azure_instance():
+            idata = gcp_instance()
```
This should be change to `azure_instance()` i guess.

tarzanek replied:
fixed

yaronkaikov commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,183 @@ def user_data(self):
+    ROOT = "root"
+    GETTING_STARTED_URL = "http://www.scylladb.com/doc/getting-started-azure/"
+    META_DATA_BASE_URL = "http://169.254.169.254/1.0/meta-data/instance/"
+    ENDPOINT_SNITCH = "GossipingPropertyFileSnitch"
```
Do we have the property file? in https://github.com/scylladb/scylla/pull/8361/files#diff-9fe4dbe99d9ee03a478c9b25175e06bbd4cef9303c0ab1b9bdeea6e52d284cd4R141 we use simpleSnitch which doesn't require one

tarzanek replied:
this is a very good comment
so we will need to add Azure Snitch to Scylla too!

tarzanek replied:
I don't think we shall use SimpleSnitch, it's not production ready snitch
so until Azure Snitch is in Scylla whoever uses this snitch later in image, will need to create the property file

yaronkaikov replied:
@tarzanek There is a template file already in Scylla repo - `conf/cassandra-rackdc.properties` - so we can actually edit this file with the required information

yaronkaikov commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,183 @@ def user_data(self):
+                    return True
+        return False
+
+    API_VERSION="?api-version=2020-06-01"
```
I think API latest version is 2021-01-01, maybe we should use it?

tarzanek replied:
yes, we can definitely upgrade, since we use the basic common functionality, updating

tarzanek commented:
thank you for review @yaronkaikov ! :-)
So with the Snitch it's a prob, I wouldn't use SimpleSnitch
is it possible image will generate the property file?
What GCP+EC2 are doing is just getting the region + zone and then using it as dc and rack , there is eventually a postifx/suffix added ( https://docs.scylladb.com/operating-scylla/system-configuration/snitch/#googlecloudsnitch )
so I guess we can add methods to get this info out of metadata for Azure and you will use it later on, how does that sound until we have an AzureSnitch in scylla cpp code base?

tarzanek commented:
e.g. sample GCP cluster with GCP snitch, suffix is _eu:
region is europe_west1
zones are b,c,d

$ nodetool status
Datacenter: europe-west1_eu
===========================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  192.168.1.208  5.63 TB    256          ?       a787247a-69c3-42f3-c55f-da53135cd789  b
UN  192.168.1.209  5.54 TB    256          ?       cbe1fbb1-c309-4375-de0d-dfc7ae03d665  c
UN  192.168.1.210  5 TB       256          ?       4bed803b-ac5b-4722-d86c-d1f9ddb6fbef  d

tarzanek commented:
in my notes I have
```
# curl -H Metadata:true --noproxy "*" "http://169.254.169.254/metadata/instance/compute?api-version=2017-04-02"
# {"location":"eastus2","name":"mlxperfTest","offer":"CentOS","osType":"Linux","platformFaultDomain":"0","platformUpdateDomain":"0","publisher":"OpenLogic","sku":"8_1-gen2","version":"8.1.2020062401","vmId":"ebacd2f3-58ec-41ea-952e-8c7d51f84abf","vmSize":"Standard_L8s_v2"}[
```

so location can be used, however I am not sure how Azure is picking zones in the region now

tarzanek commented:
once the zone logic is known, it would be easy to add it to cpp similar to https://github.com/scylladb/scylla/blob/master/locator/gce_snitch.hh

(looking at https://github.com/apache/cassandra/tree/trunk/src/java/org/apache/cassandra/locator they don't have an Azure snitch)

tarzanek commented:
but then I found https://issues.apache.org/jira/browse/CASSANDRA-9387
and he does use the zone:

https://github.com/yoshw/cassandra/commit/7101f3ad8f622612248a866fca37a8864f51943c

yaronkaikov commented:
> once the zone logic is known, it would be easy to add it to cpp similar to https://github.com/scylladb/scylla/blob/master/locator/gce_snitch.hh
> 
> (looking at https://github.com/apache/cassandra/tree/trunk/src/java/org/apache/cassandra/locator they don't have an Azure snitch)

@tarzanek when running the command `curl -H Metadata:true --noproxy "*" "http://169.254.169.254/metadata/instance?api-version=2020-06-01"`
I get the output 
```
{"compute":{"azEnvironment":"AzurePublicCloud","customData":"","evictionPolicy":"","isHostCompatibilityLayerVm":"false","licenseType":"","location":"eastus","name":"test","offer":"","osProfile":{"adminUsername":"scyllaadm","computerName":"test","disablePasswordAuthentication":"true"},"osType":"Linux","placementGroupId":"","plan":{"name":"","product":"","publisher":""},"platformFaultDomain":"0","platformUpdateDomain":"0","priority":"","provider":"Microsoft.Compute","publicKeys":[{"keyData":"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQCb/uCXTsL4iH/zOfcbOhEma+9d\r\nKRE+Q6fAHMfxhMG8hWuqVxMQMcGTyFgx/hXGBaB5kGIwScR4+Tyt0OuNLu9y2AbH\r\n5u3KXQVraED4UOQO8onAMk4ID2hkW1k46VDLLL/hJJC6ffdGYYx93nEny8mlgmFL\r\nVMnVc1tWThc2lWd68WKgTEDtkCcxSgSnEkMpFOos8/swcIMVdYNNpXlFZEm6i/ln\r\no1jRV5+yYydzK9nQsExFDaFNenHc3T86rAenTfCz+xn0f4VqBwd5sakDaGqJfAAQ\r\nxLmOqGQxyBOeoPSCUb+gDkYCE0Av1vLgGBSn16vcd+MNfetEw89x+aDg+NIV4ply\r\nvBXfXVzAI6jgOqsb0ZokN2AExhA/DOjQO2TXTCw0m+GjaLJj/pSshTl8eYt3Nyil\r\niKKnR7WKVH7i9W4SwtjiQp2JjQajc0mIeVSEAhpa3h9BFszzZjGfiAiJruCQTADl\r\nWNnbN0xq558HMgHa68Z6itlW+gRv1OK1mn25+XM= generated-by-azure\r\n","path":"/home/scyllaadm/.ssh/authorized_keys"}],"publisher":"","resourceGroupName":"scylla-images","resourceId":"/subscriptions/6c268694-47ab-43ab-b306-3c5514bc4112/resourceGroups/scylla-images/providers/Microsoft.Compute/virtualMachines/test","securityProfile":{"secureBootEnabled":"false","virtualTpmEnabled":"false"},"sku":"","storageProfile":{"dataDisks":[],"imageReference":{"id":"/subscriptions/6c268694-47ab-43ab-b306-3c5514bc4112/resourceGroups/scylla-images/providers/Microsoft.Compute/images/ScyllaDB-4.6.dev-0.20210420.dc7beec382-1","offer":"","publisher":"","sku":"","version":""},"osDisk":{"caching":"ReadWrite","createOption":"FromImage","diffDiskSettings":{"option":""},"diskSizeGB":"30","encryptionSettings":{"enabled":"false"},"image":{"uri":""},"managedDisk":{"id":"/subscriptions/6c268694-47ab-43ab-b306-3c5514bc4112/resourceGroups/scylla-images/providers/Microsoft.Compute/disks/test_disk1_300fb68b293643d6876dc8da1c1b11e6","storageAccountType":"Premium_LRS"},"name":"test_disk1_300fb68b293643d6876dc8da1c1b11e6","osType":"Linux","vhd":{"uri":""},"writeAcceleratorEnabled":"false"}},"subscriptionId":"6c268694-47ab-43ab-b306-3c5514bc4112","tags":"","tagsList":[],"userData":"","version":"","vmId":"65260cdc-7abd-4540-87c2-37972c3d2a62","vmScaleSetName":"","vmSize":"Standard_L8s_v2","zone":"1"},"network":{"interface":[{"ipv4":{"ipAddress":[{"privateIpAddress":"10.1.0.42","publicIpAddress":""}],"subnet":[{"address":"10.1.0.0","prefix":"24"}]},"ipv6":{"ipAddress":[]},"macAddress":"0022481E11CB"}]}}
```
We can use **"location":"eastus"** and **"zone":"1"**

yaronkaikov commented:
@tarzanek Please add a fixes tag to https://github.com/scylladb/scylla/issues/8325

tarzanek commented:
added both methods for location and zone and tag into the commit

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,199 @@ def user_data(self):
+            os.close(fd)
+
+    # as per https://docs.microsoft.com/en-us/azure/virtual-machines/lsv2-series
+    AZURE_NVME_DISK_SIZE_2020=1966
```
Can't we query it from the disk itself?

tarzanek replied:
we do query it, this is a sanity check that we're not below this baseline with detected nvme disk

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,199 @@ def user_data(self):
+            if af == socket.AF_INET:
+                addr, port = sa
+                if addr == "169.254.169.254":
+                    return True
```
Why this check? It seems unnecessary, and potentially breaking for someone trying to emulate the API.

tarzanek replied:
this is a good question
the idea if I remember it right was to make sure it's an internet socket and the IP resolved is the one we expect (and that is set by MS Azure folks (and it basically the same for all clouds, but dns name is different))

how do you expect this to be @avikivity to fit your idea? (since I don't feel like relaxing it, resp. don't know how to relax this check in smarter way, sorry :-( )

tarzanek replied:
note that this DNS is in local /etc/hosts , so the request shouldn't really involve network

avikivity replied:
By relaxing the check I meant removing the whole thing. Why do you care if it's an inet address or what the address is? If you can connect to that dns name, that's enough.

tarzanek replied:
okay, removing it then

avikivity replied:
It doesn't look removed.

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,199 @@ def user_data(self):
+
+        root_dev = root_dev_candidates[0].device
+        # if root_dev.startswith("/dev/mapper"):
+        #     raise Exception("mapper used for root, not checking if nvme is used ".format(root_dev))
```
What's this comment?

tarzanek replied:
removed, sorry,
I am actually tempted to replace all this method with https://github.com/scylladb/scylla/pull/8444
but it's not merged yet

tarzanek replied:
replaced it with method from 8444

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,199 @@ def user_data(self):
+    @staticmethod
+    def get_file_size_by_seek(filename):
+        "Get the file size by seeking at end"
+        fd= os.open(filename, os.O_RDONLY)
```
coding style: space before "="

tarzanek replied:
I can just drop it, since it is not needed for now

tarzanek replied:
and I fixed formatting for all code (auto-format)
yes ... again :-)

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,199 @@ def user_data(self):
+            firstDiskSize = self.get_file_size_by_seek(os.path.join("/dev/", firstDisk))
+            firstDiskSizeGB = firstDiskSize/1024/1024/1024
+            if firstDiskSizeGB >= self.AZURE_NVME_DISK_SIZE_2020:
+                self.__firstNvmeSize = firstDiskSizeGB
```
Why this check? Why not trust the disk size you measured?

tarzanek replied:
I can just drop it, since it is not needed for now

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -351,6 +351,199 @@ def user_data(self):
+            if firstDiskSizeGB >= self.AZURE_NVME_DISK_SIZE_2020:
+                self.__firstNvmeSize = firstDiskSizeGB
+            else:
+                raise Exception("First nvme is smaller than lowest expected size. ".format(firstDisk))
```
And why this check? It seems unnecessarily restrictive.

tarzanek replied:
dropping this, not needed

avikivity commented on dist/common/scripts/scylla_io_setup:
```diff
@@ -277,7 +277,54 @@ if __name__ == "__main__":
+                    disk_properties["read_iops"] = 3800000
+                    disk_properties["read_bandwidth"] = 20000 * mbs
+                    disk_properties["write_iops"] = 2546511
+                    disk_properties["write_bandwidth"] = 11998 * mbs
```
It would be nicer to make this table a dict:

{ 1: [400000, 2000, 271698, 1314],
  2: [...
}

If you want.

tarzanek replied:
can I leave this for refactoring that I promised to Bentsi?

avikivity replied:
Yes ("If you want" above).

avikivity commented on dist/common/scripts/scylla_io_setup:
```diff
@@ -277,7 +277,54 @@ if __name__ == "__main__":
+                    disk_properties["read_bandwidth"] = 20000 * mbs
+                    disk_properties["write_iops"] = 2546511
+                    disk_properties["write_bandwidth"] = 11998 * mbs
+                properties_file = open(etcdir() + "/scylla.d/io_properties.yaml", "w")
```
Let's have an `else` branch so we don't crash if we don't match anything.

tarzanek replied:
added + added for GCP too

tarzanek replied:
it needs to be added to AWS, but I'd like to defer that to a different PR

avikivity replied:
Sure (would be better to defer GCP too)

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -378,13 +378,7 @@ def is_azure_instance():
-                if addr == "169.254.169.254":
-                    return True
-        return False
+        return True
```
Oh, it's removed here. Don't introduce something in patch 1 and remove it in patch 3, unless your goal is to annoy the reviewer.

tarzanek replied:
will squash, sorry, was trying to make the change after review obvious

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -427,6 +421,80 @@ def getEphemeralOsDisks(self):
+#       },
+#       "writeAcceleratorEnabled": "false"
+#     }
+# }
```
Please explain what this comment means so somebody looking at it in 4 months doesn't have to guess.

tarzanek replied:
removing, it was just a placeholder for the refactoring

tarzanek replied:
will remove

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -518,6 +586,37 @@ def is_recommended_instance_size(self):
+    #         return os.lseek(fd, 0, os.SEEK_END)
+    #     finally:
+    #         os.close(fd)
+
```
What's this comment?

tarzanek replied:
removing, it was just a placeholder for the refactoring

tarzanek commented:
rewritten the commits, squashed the removal of check for dns reply
removed comments that didn't belong to the last commit
and removed the placeholder comments for refactoring
and since it's 4.5. I used the --force to push :-D
